### PR TITLE
make locking more fine-grained on client close

### PR DESCRIPTION
**Pull request by fearful-symmetry**
This came out of a discussion with @urso -- Client and pipeline close operations on libbeat can block depending on `WaitClose` and other settings. We could potentially block _other_ client start operations while the pipelines drained. This attempts to make the locking more fine-grained so we don't hold a mutex while queues drain.

### CODE REVIEW

File: x-pack/dockerlogbeat/pipelinemanager/libbeattools.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
+package pipelinemanager
+
+// makeConfigHash is the helper function that turns a user config into a hash
```
Comment (sayden) [inline x-pack/dockerlogbeat/pipelinemanager/libbeattools.go]: Incorrect line :sweat_smile:

File: x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go [hunk 2]
```diff
@@ -55,170 +48,138 @@ func NewPipelineManager(logCfg *common.Config) *PipelineManager {
...
+	defer pm.mu.Unlock()
+	pm.clients[clientFile] = cl
+	pm.pipelines[hashstring].refCount++
+}
 
-	outMap := make(map[string]interface{})
+// removeClient deregisters a client
+func (pm *PipelineManager) removeClient(file string) (*ClientLogger, error) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
```
Comment (sayden) [inline x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go]: I think will unlock an unlocked mutex from line 175 with a `fatal error: sync: unlock of unlocked mutex`

Comment (fearful-symmetry) [reply x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go]: Good catch!

**Comment by fearful-symmetry** [comment]
@sayden fixed!

File: x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go [hunk 3]
```diff
@@ -55,170 +48,137 @@ func NewPipelineManager(logCfg *common.Config) *PipelineManager {
...
+	//If we don't have an existing pipeline for this hash, make one
+	exists := pm.checkIfHashExists(logOptsConfig)
+	var pipeline *Pipeline
+	var err error
+	if !exists {
+		pipeline, err = loadNewPipeline(logOptsConfig, file, pm.Logger)
+		if err != nil {
+			return nil, errors.Wrap(err, "error loading pipeline")
+		}
+		pm.registerPipeline(pipeline, hashstring)
```
Comment (urso) [inline x-pack/dockerlogbeat/pipelinemanager/pipelineManager.go]: This pattern looks like a race, potentially creating zombie-pipelines. At least the if block should be protected and there should be another check (similar to double-checked locking pattern).

Does docker synchronize API calls on logging plugins? In this case we might be safe here.

**Comment by urso** [comment]
What happened to the notice file? Given the change itself I wonder why we need to update it in this PR.

**Pull request merged**